src/contract/contract.ts

```diff
@@ -26,165 +26,310 @@ import { node } from '../api'
+      for (const file of files) {
+        if (file.type === SourceType.Contract) {
+          const artifact = await Contract.fromArtifactFile(file.artifactPath)
+          contracts.push(new ArtifactInfo(file, artifact))
+        } else if (file.type === SourceType.Script) {
+          const artifact = await Script.fromArtifactFile(file.artifactPath)
+          scripts.push(new ArtifactInfo(file, artifact))
+        }
+      }
+      return new Project(files, contracts, scripts)
+    } catch (error) {
+      console.log('Failed to load artifacts, try to re-compile contracts...')
+      return Project.compile(files)
     }
   }
 
-  protected static async _from<T extends { sourceCodeSha256: string }>(
-    provider: NodeProvider,
-    sourceFile: SourceFile,
-    loadContractStr: (sourceFile: SourceFile, importsCache: string[]) => Promise<string>,
-    compile: (
-      provider: NodeProvider,
-      sourceFile: SourceFile,
-      contractStr: string,
-      contractHash: string,
-      errorOnWarnings: boolean,
-      ignoreUnusedConstantsWarnings: boolean
-    ) => Promise<T>,
-    errorOnWarnings: boolean,
-    ignoreUnusedConstantsWarnings: boolean
-  ): Promise<T> {
-    Common.checkFileNameExtension(sourceFile.contractPath)
-
-    const contractStr = await loadContractStr(sourceFile, [])
-    const contractHash = cryptojs.SHA256(contractStr).toString()
-    const existingContract = this._getArtifactFromCache(contractHash)
-    if (typeof existingContract !== 'undefined') {
-      return existingContract as unknown as T
+  private static async loadSourceFile(dirPath: string, filename: string): Promise<SourceFile> {
+    const fullPath = dirPath + '/' + filename
+    if (!filename.endsWith('.ral')) {
+      throw new Error(`Invalid filename: ${fullPath}, smart contract file name should end with ".ral"`)
+    }
+
+    const sourceBuffer = await fsPromises.readFile(fullPath)
+    const sourceStr = sourceBuffer.toString()
+    const results = this.matchers.map((m) => m.match(sourceStr))
+    const matchNumber = results.reduce((a, b) => a + b, 0)
+    if (matchNumber === 0) {
+      throw new Error(`No contract defined in file: ${fullPath}`)
+    }
+    if (matchNumber > 1) {
+      throw new Error(`Multiple definitions in file: ${fullPath}`)
+    }
+    const matcherIndex = results.indexOf(1)
+    const type = this.matchers[matcherIndex].type
+    return new SourceFile(type, dirPath, filename, sourceStr)
+  }
+
+  private static async loadSourceFiles(rootPath: string): Promise<SourceFile[]> {
+    const loadDir = async function (dirPath: string, results: SourceFile[]): Promise<void> {
+      const dirents = await fsPromises.readdir(dirPath, { withFileTypes: true })
+      for (const dirent of dirents) {
+        if (dirent.isFile()) {
+          const file = await Project.loadSourceFile(dirPath, dirent.name)
+          results.push(file)
+        } else {
+          const newPath = dirPath + '/' + dirent.name
```

Comment:
Just in case I forget: this will not work on Windows system. We will need another task to support windows

Reply:
I used the nodejs path lib at the beginning, but seems it doesn't work with the webpack, and the [path-browserify](https://github.com/browserify/path-browserify) only implements the POSIX functions.